src/tools/lifter/bir_lifting_machinesLib_instances.sml

```diff
@@ -409,4 +409,175 @@ val _ = assert bmr_rec_sanity_check (m0_bmr_rec_BigEnd_Main)
+  val sel_fl_tm = if sel_fl then T else F;
+
+  val m0_mod_step_hex = m0_mod_stepLib.thumb_mod_step_hex (endian_fl, sel_fl);
+
+  val next_state_tm = (prim_mk_const{Name="NextStateM0_mod", Thy="m0_mod_step"});
+
+  val simp_conv = (SIMP_CONV (arith_ss++bitstringLib.v2w_n2w_ss++wordsLib.SIZES_ss)
+     ((if endian_fl then m0_extra_FOLDS_BE else m0_extra_FOLDS_LE)::[nzcv_FOLDS_M0,
+     EQ_13w_EVAL, EQ_15w_EVAL, R_name_EVAL, bir_auxiliaryTheory.w2w_n2w,
+     m0_extra_FOLDS_GEN, Mode_Handler_INTRO, bir_auxiliaryTheory.align_aligned_add,
+     bir_auxiliaryTheory.align_aligned_sub, LowestSetBit_n2w, numeral_bitTheory.LOWEST_SET_BIT,
+     alignmentTheory.aligned_numeric, alignmentTheory.align_aligned, wordsTheory.word_bit_n2w]));
+
+  val compset_2 = reduceLib.num_compset ();
+  val _ = bitLib.add_bit_compset compset_2
+
+  val simp_conv2 = SIMP_CONV (arith_ss++wordsLib.WORD_ARITH_ss ++ wordsLib.WORD_LOGIC_ss++wordsLib.SIZES_ss) [wordsTheory.n2w_11, m0_extra_FOLDS_GEN, wordsTheory.word_msb, wordsTheory.word_bit_n2w] THENC
+                   (SIMP_CONV std_ss [word_add_to_sub_TYPES, wordsTheory.WORD_SUB_INTRO, alignmentTheory.aligned_numeric, wordsTheory.WORD_MULT_CLAUSES] THENC
+                    computeLib.CBV_CONV compset_2);
+
+  val bmr_extra_M0_mod' = REWRITE_RULE [] (SPECL [endian_fl_tm, sel_fl_tm] bmr_extra_M0_mod)
+  fun m0_mod_extra_THMS vn = let
+     val thm0  = SPEC vn bmr_extra_M0_mod'
+     val thm1a = ASSUME (lhs (concl thm0))
+     val thm1 = CONV_RULE (K thm0) thm1a
+  in
+     CONJUNCTS thm1
+  end;
+
+
+  val bmr_ms_mem_contains_M0_mod_2' = SPECL [endian_fl_tm, sel_fl_tm] bmr_ms_mem_contains_M0_mod_2
+  val bmr_ms_mem_contains_M0_mod_4' = SPECL [endian_fl_tm, sel_fl_tm] bmr_ms_mem_contains_M0_mod_4
+
+  fun prepare_mem_contains_thms vn hex_code =
+  let
+    val bytes = m0_mod_reorder_bytes endian_fl (bytes_of_hex_code hex_code)
+    val ms_thm = if (length bytes = 2) then
+                    bmr_ms_mem_contains_M0_mod_2'
+                 else if (length bytes = 4) then
+                    bmr_ms_mem_contains_M0_mod_4'
+                 else failwith "invalid hex-code";
+
+    val thm0 = SPECL (vn :: bytes) ms_thm
+
+    val thm1a = ASSUME (lhs (concl thm0))
+    val thm2 = CONV_RULE (K thm0) thm1a
+  in
+    CONJUNCTS thm2
+  end;
+
+(* val thm = hd step_thms0  *)
+   fun process_m0_mod_thm vn pc_mem_thms thm = let
+     val thm0 = bmr_normalise_step_thm next_state_tm vn thm
+     val thm1 = HYP_CONV_RULE (K true) simp_conv thm0
+     val thm2 = foldl (fn (pre_thm, thm) => PROVE_HYP pre_thm thm) thm1
+       (pc_mem_thms @ (m0_mod_extra_THMS vn))
+
+     val thm3 = DISCH_ALL thm2
+     val thm4 = CONV_RULE (simp_conv THENC simp_conv2) thm3
+     val thm5 = UNDISCH_ALL thm4
+
+     (* check that theorem structure did not get destroyed by e.g.
+        contradicting assumptions. *)
+     val _ = dest_eq (concl thm5) handle HOL_ERR _ => failwith "m0_mod_step_hex': unexpected resulting theorem"
```

Comment:
(I) Prefer `wrap_exn` to `handle HOL_ERR _ => failwith`.

Reply:
ok, we can do this